src/parser/toplevel.lisp

```diff
@@ -554,56 +536,213 @@ consume all attributes"))))
+           (package (make-instance 'toplevel-package
+                      :name (symbol-name package-name)
+                      :docstring (consume-string iter :optional t))))
+      (loop :while (next-value-p iter)
+            :do (parse-package-clause package (consume-value iter) file))
+      package)))
 
-  ;; Package declarations cannot contain more than two forms
-  (when (cst:consp (cst:rest (cst:rest form)))
-    (error 'parse-error
-           :err (se:source-error
-                 :span (cst:source (cst:first (cst:rest (cst:rest form))))
-                 :file file
-                 :message "Malformed package declaration"
-                 :primary-note "unexpected forms")))
-
-  (unless (identifierp (cst:raw (cst:second form)))
-    (error 'parse-error
-           :err (se:source-error
-                 :span (cst:source (cst:second form))
-                 :file file
-                 :message "Malformed package declaration"
-                 :primary-note "package name must be a symbol")))
-
-  (let* ((package-name (symbol-name (cst:raw (cst:second form))))
-
-         (package (find-package package-name)))
-
-    (unless package
-      (setf package (make-package package-name :use '("COALTON" "COALTON-PRELUDE"))))
+;; Empty package for reading (package) forms
 
-    package))
+(defpackage #:coalton-impl/parser/%defpackage
+  (:documentation "The sole purpose of this package is to provide storage for symbols that are interned while Coalton package forms are read.")
+  (:use))
 
+(defun read-toplevel-package (stream file)
+  "Read and parse a Coalton toplevel package form."
+  (let ((*package* (find-package 'coalton-impl/parser/%defpackage)))
+    (unwind-protect
+         (multiple-value-bind (form presentp)
+             (maybe-read-form stream *coalton-eclector-client*)
+           (unless presentp
+             (error 'parse-error
+                    :err (se:source-error
+                          :span (cons (- (file-position stream) 2)
+                                      (- (file-position stream) 1))
+                          :file file
+                          :message "Unexpected EOF"
+                          :primary-note "missing package form")))
+           (parse-package form file))
+      (do-symbols (symbol)
+        (unintern symbol)))))
+
+(defun make-defpackage (toplevel-package)
+  "Generate a Lisp defpackage form from a Caolton toplevel-package structure."
+  (with-slots (name import import-from import-as export) toplevel-package
+    `(defpackage ,name
+       (:use "COALTON" ,@import)
+       ,@(mapcar (lambda (form)
+                   `(:import-form ,@form))
+                 import-from)
+       (:local-nicknames ,@import-as)
+       (:export ,@export))))
+
+(defun lisp-package (toplevel-package)
+  "Create a Lisp package by evlauating the defpackage representation of a TOPLEVEL-PACKAGE form."
+  (eval (make-defpackage toplevel-package)))
```

Comment:
If constructing the package "by hand" using `make-package` etc is too onerous, this is an acceptable alternative. 

I'd probably use `make-package` though, as it allows for a greater degree of control and doesn't call up the whole evaluator.

Reply:
That was the first thing I explored, and the main thing with that was, there are a number of special mutation functions (e.g., those for package local nicknames) that have varying forms in the supported Lisp implementations. There was also a remark here about potentially diverging behavior, so the eval is a cheap and effective way of getting consistency between coalton-parse and lisp-load time package definitions, and retaining CL package redefinition semantics (for now).